[CI] GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoHex failure

### CI Link

https://gradle-enterprise.elastic.co/s/j3ajt4vzqybxs

### Repro line

gradlew ':x-pack:plugin:spatial:internalClusterTest' --tests "org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoHex" -Dtests.seed=67641D5B7DA4755B -Dtests.locale=en-IN -Dtests.timezone=Canada/Saskatchewan -Druntime.java=17 -Dtests.fips.enabled=true

### Does it reproduce?

No

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

The test in CI link above doesn't repro for me but I did run the test with multiple iterations. After around 250 of them, one failed for me:

```
REPRODUCE WITH: gradlew ':x-pack:plugin:spatial:internalClusterTest' --tests "org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoHex {seed=[3478FF74E8D66BB6:A2F98EBD906145BB]}" -Dtests.seed=3478FF74E8D66BB6 -Dtests.locale=sr-RS -Dtests.timezone=Europe/Luxembourg -Druntime.java=19

org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT > testGeoShapeGeoHex {seed=[3478FF74E8D66BB6:A2F98EBD906145BB]} FAILED
    java.lang.AssertionError: Expected hits at precision 15
    Expected: <3L>
         but: was <4L>
        at __randomizedtesting.SeedInfo.seed([3478FF74E8D66BB6:A2F98EBD906145BB]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.assertQuery(GeoGridAggAndQueryConsistencyIT.java:224)
        at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.doTestGrid(GeoGridAggAndQueryConsistencyIT.java:215)
        at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.doTestGeohexGrid(GeoGridAggAndQueryConsistencyIT.java:124)
        at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoHex(GeoGridAggAndQueryConsistencyIT.java:92)
```

--- a/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/GeoGridAggAndQueryConsistencyIT.java
+++ b/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/GeoGridAggAndQueryConsistencyIT.java
@@ -137,6 +137,53 @@ public class GeoGridAggAndQueryConsistencyIT extends ESIntegTestCase {
         }
     }
 
+    public void testKnownIssueWithCellIntersectingPolygonAndBoundingBox() throws IOException {
+        XContentBuilder xcb = XContentFactory.jsonBuilder()
+            .startObject()
+            .startObject("properties")
+            .startObject("geometry")
+            .field("type", "geo_shape")
+            .endObject()
+            .endObject()
+            .endObject();
+        client().admin().indices().prepareCreate("test").setMapping(xcb).get();
+
+        BulkRequestBuilder builder = client().prepareBulk();
+        builder.add(
+            new IndexRequest("test").source("{\"geometry\" : \"POINT (169.12088680200193 86.17678739494652)\"}", XContentType.JSON)
+        );
+        builder.add(
+            new IndexRequest("test").source("{\"geometry\" : \"POINT (169.12088680200193 86.17678739494652)\"}", XContentType.JSON)
+        );
+        String mp = "POLYGON ((150.0 70.0, 150.0 85.91811374669217, 168.77544806565834 85.91811374669217, 150.0 70.0))";
+        builder.add(new IndexRequest("test").source("{\"geometry\" : \"" + mp + "\"}", XContentType.JSON));
+
+        assertFalse(builder.get().hasFailures());
+        client().admin().indices().prepareRefresh("test").get();
+
+        // BBOX (172.21916569181505, -173.17785081207947, 86.17678739494652, 83.01600086049713)
+        GeoBoundingBox boundingBox = new GeoBoundingBox(
+            new GeoPoint(86.17678739494652, 172.21916569181505),
+            new GeoPoint(83.01600086049713, 179)
+        );
+        int precision = 4;
+        GeoGridAggregationBuilder builderPoint = new GeoHexGridAggregationBuilder("geometry").field("geometry")
+            .precision(precision)
+            .setGeoBoundingBox(boundingBox)
+            .size(256 * 256);
+        SearchResponse response = client().prepareSearch("test").addAggregation(builderPoint).setSize(0).get();
+        InternalGeoGrid<?> gridPoint = response.getAggregations().get("geometry");
+        for (InternalGeoGridBucket bucket : gridPoint.getBuckets()) {
+            assertThat(bucket.getDocCount(), Matchers.greaterThan(0L));
+            QueryBuilder queryBuilder = new GeoGridQueryBuilder("geometry").setGridId(
+                GeoGridQueryBuilder.Grid.GEOHEX,
+                bucket.getKeyAsString()
+            );
+            response = client().prepareSearch("test").setTrackTotalHits(true).setQuery(queryBuilder).get();
+            assertThat(response.getHits().getTotalHits().value, Matchers.equalTo(bucket.getDocCount()));
+        }
+    }
+
     private void doTestGeohashGrid(String fieldType, Supplier<Geometry> randomGeometriesSupplier) throws IOException {
         doTestGrid(
             1,
@@ -267,7 +314,7 @@ public class GeoGridAggAndQueryConsistencyIT extends ESIntegTestCase {
             QueryBuilder queryBuilder = queryFunction.apply("geometry", bucket.getKeyAsString());
             SearchResponse response = client().prepareSearch("test").setTrackTotalHits(true).setQuery(queryBuilder).get();
             assertThat(
-                "Expected hits at precision " + precision,
+                "Expected hits at precision " + precision + " for H3 cell " + bucket.getKeyAsString(),
                 response.getHits().getTotalHits().value,
                 Matchers.equalTo(bucket.getDocCount())
             );

--- a/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/GeoGridAggAndQueryConsistencyIT.java
+++ b/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/GeoGridAggAndQueryConsistencyIT.java
@@ -88,9 +88,53 @@ public class GeoGridAggAndQueryConsistencyIT extends ESIntegTestCase {
         );
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/92551")
     public void testGeoShapeGeoHex() throws IOException {
         doTestGeohexGrid(GeoShapeWithDocValuesFieldMapper.CONTENT_TYPE, () -> GeometryTestUtils.randomGeometryWithoutCircle(0, false));
+    }
+
+    public void testKnownIssueWithCellLeftOfDatelineTouchingPolygonOnRightOfDateline() throws IOException {
+        XContentBuilder xcb = XContentFactory.jsonBuilder()
+            .startObject()
+            .startObject("properties")
+            .startObject("geometry")
+            .field("type", "geo_shape")
+            .endObject()
+            .endObject()
+            .endObject();
+        client().admin().indices().prepareCreate("test").setMapping(xcb).get();
+
+        BulkRequestBuilder builder = client().prepareBulk();
+        builder.add(
+            new IndexRequest("test").source("{\"geometry\" : \"BBOX (179.99999, 180.0, -11.29550, -11.29552)\"}", XContentType.JSON)
+        );
+        builder.add(
+            new IndexRequest("test").source("{\"geometry\" : \"BBOX (-180.0, -179.99999, -11.29550, -11.29552)\"}", XContentType.JSON)
+        );
+
+        assertFalse(builder.get().hasFailures());
+        client().admin().indices().prepareRefresh("test").get();
+
+        GeoBoundingBox boundingBox = new GeoBoundingBox(new GeoPoint(-11.29550, 179.999992), new GeoPoint(-11.29552, -179.999992));
+
+        GeoGridAggregationBuilder builderPoint = new GeoHexGridAggregationBuilder("geometry").field("geometry")
+            .precision(15)
+            .setGeoBoundingBox(boundingBox)
+            .size(256 * 256);
+        SearchResponse response = client().prepareSearch("test").addAggregation(builderPoint).setSize(0).get();
+        InternalGeoGrid<?> gridPoint = response.getAggregations().get("geometry");
+        for (InternalGeoGridBucket bucket : gridPoint.getBuckets()) {
+            assertThat(bucket.getDocCount(), Matchers.greaterThan(0L));
+            QueryBuilder queryBuilder = new GeoGridQueryBuilder("geometry").setGridId(
+                GeoGridQueryBuilder.Grid.GEOHEX,
+                bucket.getKeyAsString()
+            );
+            response = client().prepareSearch("test").setTrackTotalHits(true).setQuery(queryBuilder).get();
+            assertThat(
+                "Bucket " + bucket.getKeyAsString(),
+                response.getHits().getTotalHits().value,
+                Matchers.equalTo(bucket.getDocCount())
+            );
+        }
     }
 
     private void doTestGeohashGrid(String fieldType, Supplier<Geometry> randomGeometriesSupplier) throws IOException {

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3CartesianGeometry.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3CartesianGeometry.java
@@ -109,14 +109,13 @@ class H3CartesianGeometry extends LatLonGeometry {
             if (Component2D.within(getMinX(), getMaxX(), getMinY(), getMaxY(), minX, maxX, minY, maxY)) {
                 return PointValues.Relation.CELL_CROSSES_QUERY;
             }
-            int numCorners = numberOfCorners(minX, maxX, minY, maxY);
-            if (numCorners == 4) {
-                // need to check in case we are crossing the dateline
-                if (crossesDateline && H3CartesianUtil.crossesBox(xs, ys, xs.length, crossesDateline, minX, maxX, minY, maxY, true)) {
+            ContainsCorners containsCorners = containsCorners(minX, maxX, minY, maxY);
+            if (containsCorners == ContainsCorners.ALL) {
+                if (H3CartesianUtil.crossesBox(xs, ys, xs.length, crossesDateline, minX, maxX, minY, maxY, true)) {
                     return PointValues.Relation.CELL_CROSSES_QUERY;
                 }
                 return PointValues.Relation.CELL_INSIDE_QUERY;
-            } else if (numCorners == 0) {
+            } else if (containsCorners == ContainsCorners.NONE) {
                 if (Component2D.containsPoint(xs[0], ys[0], minX, maxX, minY, maxY)) {
                     return PointValues.Relation.CELL_CROSSES_QUERY;
                 }
@@ -325,7 +324,7 @@ class H3CartesianGeometry extends LatLonGeometry {
             return H3CartesianUtil.crossesLine(xs, ys, xs.length, crossesDateline, minX, maxX, minY, maxY, aX, aY, bX, bY, includeBoundary);
         }
 
-        private int numberOfCorners(double minX, double maxX, double minY, double maxY) {
+        private ContainsCorners containsCorners(double minX, double maxX, double minY, double maxY) {
             int containsCount = 0;
             if (contains(minX, minY)) {
                 containsCount++;
@@ -334,18 +333,24 @@ class H3CartesianGeometry extends LatLonGeometry {
                 containsCount++;
             }
             if (containsCount == 1) {
-                return containsCount;
+                return ContainsCorners.SOME;
             }
             if (contains(maxX, maxY)) {
                 containsCount++;
             }
             if (containsCount == 2) {
-                return containsCount;
+                return ContainsCorners.SOME;
             }
             if (contains(minX, maxY)) {
                 containsCount++;
             }
-            return containsCount;
+            return containsCount == 0 ? ContainsCorners.NONE : containsCount == 4 ? ContainsCorners.ALL : ContainsCorners.SOME;
+        }
+
+        private enum ContainsCorners {
+            NONE,
+            SOME,
+            ALL
         }
     }
 }

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitor.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitor.java
@@ -297,21 +297,21 @@ class GeoHexVisitor extends TriangleTreeVisitor.TriangleTreeDecodedVisitor {
         }
 
         boolean within = false;
-        if (edgeIntersectsQuery(aX, aY, bX, bY, false)) {
+        if (edgeIntersectsQuery(aX, aY, bX, bY, ab)) {
             if (ab) {
                 return GeoRelation.QUERY_CROSSES;
             }
             within = true;
         }
 
-        if (edgeIntersectsQuery(bX, bY, cX, cY, false)) {
+        if (edgeIntersectsQuery(bX, bY, cX, cY, bc)) {
             if (bc) {
                 return GeoRelation.QUERY_CROSSES;
             }
             within = true;
         }
 
-        if (edgeIntersectsQuery(cX, cY, aX, aY, false)) {
+        if (edgeIntersectsQuery(cX, cY, aX, aY, ca)) {
             if (ca) {
                 return GeoRelation.QUERY_CROSSES;
             }

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitor.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitor.java
@@ -234,7 +234,7 @@ class GeoHexVisitor extends TriangleTreeVisitor.TriangleTreeDecodedVisitor {
             return GeoRelation.QUERY_CONTAINS;
         }
         // 2. check crossings
-        if (edgeIntersectsQuery(aX, aY, bX, bY, true)) {
+        if (edgeIntersectsQuery(aX, aY, bX, bY)) {
             return GeoRelation.QUERY_CROSSES;
         }
         return GeoRelation.QUERY_DISJOINT;
@@ -297,21 +297,21 @@ class GeoHexVisitor extends TriangleTreeVisitor.TriangleTreeDecodedVisitor {
         }
 
         boolean within = false;
-        if (edgeIntersectsQuery(aX, aY, bX, bY, ab)) {
+        if (edgeIntersectsQuery(aX, aY, bX, bY)) {
             if (ab) {
                 return GeoRelation.QUERY_CROSSES;
             }
             within = true;
         }
 
-        if (edgeIntersectsQuery(bX, bY, cX, cY, bc)) {
+        if (edgeIntersectsQuery(bX, bY, cX, cY)) {
             if (bc) {
                 return GeoRelation.QUERY_CROSSES;
             }
             within = true;
         }
 
-        if (edgeIntersectsQuery(cX, cY, aX, aY, ca)) {
+        if (edgeIntersectsQuery(cX, cY, aX, aY)) {
             if (ca) {
                 return GeoRelation.QUERY_CROSSES;
             }
@@ -328,13 +328,13 @@ class GeoHexVisitor extends TriangleTreeVisitor.TriangleTreeDecodedVisitor {
     /**
      * returns true if the edge (defined by (ax, ay) (bx, by)) intersects the query
      */
-    private boolean edgeIntersectsQuery(double ax, double ay, double bx, double by, boolean includeBoundary) {
+    private boolean edgeIntersectsQuery(double ax, double ay, double bx, double by) {
         final double minX = Math.min(ax, bx);
         final double maxX = Math.max(ax, bx);
         final double minY = Math.min(ay, by);
         final double maxY = Math.max(ay, by);
         return boxesAreDisjoint(minX, maxX, minY, maxY) == false
-            && H3CartesianUtil.crossesLine(xs, ys, numPoints, crossesDateline, minX, maxX, minY, maxY, ax, ay, bx, by, includeBoundary);
+            && H3CartesianUtil.crossesLine(xs, ys, numPoints, crossesDateline, minX, maxX, minY, maxY, ax, ay, bx, by, true);
     }
 
     /**

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3CartesianGeometryTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3CartesianGeometryTests.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.common;
+
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.geo.LatLonGeometry;
+import org.apache.lucene.index.PointValues;
+import org.elasticsearch.h3.H3;
+import org.elasticsearch.test.ESTestCase;
+
+public class H3CartesianGeometryTests extends ESTestCase {
+    public void testPolarH3Crossing() {
+        long h3 = H3.geoToH3(-90, 0, 2);
+        Component2D component2D = LatLonGeometry.create(H3CartesianUtil.getLatLonGeometry(h3));
+        assertEquals(
+            PointValues.Relation.CELL_CROSSES_QUERY,
+            component2D.relate(-155.15317029319704, 163.99789148010314, -88.69673718698323, -88.26949733309448)
+        );
+    }
+}

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
@@ -200,6 +200,52 @@ public class GeoHexTilerTests extends GeoGridTilerTestCase {
         assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
     }
 
+    public void testTroublesomeCellLevel2_BoundedGeoShapeCellValues() throws Exception {
+        int precision = 2;
+        String wkt = """
+            GEOMETRYCOLLECTION (
+              GEOMETRYCOLLECTION (
+                POINT(-170 0),
+                POINT (-178.5 0)
+              )
+            )
+            """;
+        GeoBoundingBox boundingBox = new GeoBoundingBox(new GeoPoint(4E-4, 179.999), new GeoPoint(-4E-4, -179.999));
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, wkt);
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues cellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getBoundedGridTiler(boundingBox, precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(cellValues.advanceExact(0));
+        int numBuckets = cellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, boundingBox);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
+    }
+
+    public void testTroublesomeCellLevel4_BoundedGeoShapeCellValues() throws Exception {
+        int precision = 4;
+        String polygon = "POLYGON ((150.0 70.0, 150.0 85.91811374669217, 168.77544806565834 85.91811374669217, 150.0 70.0))";
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
+        GeoBoundingBox boundingBox = new GeoBoundingBox(
+            new GeoPoint(86.17678739494652, 172.21916569181505),
+            new GeoPoint(83.01600086049713, 179)
+        );
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues cellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getBoundedGridTiler(boundingBox, precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(cellValues.advanceExact(0));
+        int numBuckets = cellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, boundingBox);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
+    }
+
     private void assertCorner(long[] h3bins, Point point, int precision, String msg) throws IOException {
         GeoShapeValues.GeoShapeValue cornerValue = geoShapeValue(point);
         GeoShapeCellValues cornerValues = new GeoShapeCellValues(
@@ -260,37 +306,40 @@ public class GeoHexTilerTests extends GeoGridTilerTestCase {
 
     @Override
     protected int expectedBuckets(GeoShapeValues.GeoShapeValue geoValue, int precision, GeoBoundingBox bbox) throws Exception {
-        return computeBuckets(H3.getLongRes0Cells(), bbox, geoValue, precision);
+        BoundedGeoHexGridTiler bounded = bbox == null ? null : new BoundedGeoHexGridTiler(precision, bbox);
+        UnboundedGeoHexGridTiler predicate = new UnboundedGeoHexGridTiler(precision);
+        return computeBuckets(H3.getLongRes0Cells(), bounded, predicate, geoValue, precision);
     }
 
-    private int computeBuckets(long[] children, GeoBoundingBox bbox, GeoShapeValues.GeoShapeValue geoValue, int finalPrecision)
-        throws IOException {
+    private int computeBuckets(
+        long[] children,
+        BoundedGeoHexGridTiler bounded,
+        UnboundedGeoHexGridTiler predicate,
+        GeoShapeValues.GeoShapeValue geoValue,
+        int finalPrecision
+    ) throws IOException {
         int count = 0;
         for (long child : children) {
             if (H3.getResolution(child) == finalPrecision) {
-                if (intersects(child, geoValue, bbox, finalPrecision)) {
+                if (intersects(child, geoValue, bounded, predicate)) {
                     count++;
                 }
             } else {
-                count += computeBuckets(H3.h3ToChildren(child), bbox, geoValue, finalPrecision);
+                count += computeBuckets(H3.h3ToChildren(child), bounded, predicate, geoValue, finalPrecision);
             }
         }
         return count;
     }
 
-    private boolean intersects(long h3, GeoShapeValues.GeoShapeValue geoValue, GeoBoundingBox bbox, int finalPrecision) throws IOException {
-        if (addressIntersectsBounds(h3, bbox, finalPrecision) == false) {
+    private boolean intersects(
+        long h3,
+        GeoShapeValues.GeoShapeValue geoValue,
+        BoundedGeoHexGridTiler bounded,
+        UnboundedGeoHexGridTiler predicate
+    ) throws IOException {
+        if (bounded != null && bounded.h3IntersectsBounds(h3) == false) {
             return false;
         }
-        UnboundedGeoHexGridTiler predicate = new UnboundedGeoHexGridTiler(finalPrecision);
         return predicate.relateTile(geoValue, h3) != GeoRelation.QUERY_DISJOINT;
-    }
-
-    private boolean addressIntersectsBounds(long h3, GeoBoundingBox bbox, int finalPrecision) {
-        if (bbox == null) {
-            return true;
-        }
-        BoundedGeoHexGridTiler predicate = new BoundedGeoHexGridTiler(finalPrecision, bbox);
-        return predicate.h3IntersectsBounds(h3);
     }
 }

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitorTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexVisitorTests.java
@@ -8,6 +8,8 @@
 package org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid;
 
 import org.apache.lucene.tests.geo.GeoTestUtil;
+import org.elasticsearch.common.geo.GeoBoundingBox;
+import org.elasticsearch.common.geo.GeoPoint;
 import org.elasticsearch.common.geo.GeometryNormalizer;
 import org.elasticsearch.common.geo.Orientation;
 import org.elasticsearch.geometry.Geometry;
@@ -194,5 +196,39 @@ public class GeoHexVisitorTests extends ESTestCase {
             reader.visit(visitor);
             assertEquals(GeoRelation.QUERY_CROSSES, visitor.relation());
         }
+    }
+
+    // Testing issue with a specific cell left of the dateline that touches the dateline with one point
+    // Intersecting a polygon on the other side of the dateline.
+    public void testSpecificCellTouchesDateline() throws IOException {
+        long h3 = 646728346019944298L;
+        GeoBoundingBox box = new GeoBoundingBox(new GeoPoint(-11.29550, -180), new GeoPoint(-11.29552, -179.99999));
+        GeoHexVisitor visitor = new GeoHexVisitor();
+        visitor.reset(h3);
+        {
+            // Polygon on the same side of the dateline (overlaps)
+            Polygon polygon = makePolygonFromBox(179.99999, 180, -11.29550, -11.29552);
+            GeometryDocValueReader reader = GeoTestUtils.geometryDocValueReader(polygon, CoordinateEncoder.GEO);
+            reader.visit(visitor);
+            assertEquals("Polygon on the same side of the dateline", GeoRelation.QUERY_CROSSES, visitor.relation());
+        }
+        {
+            // Polygon on the other side of the dateline (just touches)
+            Polygon polygon = makePolygonFromBox(-180, -179.99999, -11.29550, -11.29552);
+            GeometryDocValueReader reader = GeoTestUtils.geometryDocValueReader(polygon, CoordinateEncoder.GEO);
+            reader.visit(visitor);
+            assertEquals(
+                "Polygon on the other side of the dateline (touches at single point)",
+                GeoRelation.QUERY_CROSSES,
+                visitor.relation()
+            );
+        }
+    }
+
+    private Polygon makePolygonFromBox(double left, double right, double top, double bottom) {
+        double[] x = new double[] { left, right, right, left, left };
+        double[] y = new double[] { bottom, bottom, top, top, bottom };
+        LinearRing ring = new LinearRing(x, y);
+        return new Polygon(ring);
     }
 }

